core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java

```diff
@@ -225,29 +217,35 @@ private void constructorArg(int position, ParseField parseField, Object value) {
             }
             constructorArgs[position] = value;
             constructorArgsCollected++;
-            if (constructorArgsCollected == constructorArgNames.size()) {
-                try {
-                    unspool();
-                } catch (ParsingException e) {
-                    throw new ParsingException(e.getLineNumber(), e.getColumnNumber(),
-                            "failed to build [" + objectParser.getName() + "] after last required field arrived", e);
-                } catch (Exception e) {
-                    throw new ParsingException(null, "Failed to build [" + objectParser.getName() + "] after last required field arrived",
-                            e);
+            if (constructorArgsCollected != constructorArgNames.size()) {
+                return;
+            }
+            try {
+                targetObject = builder.apply(constructorArgs);
+                while (spooledFields > 0) {
+                    spooledFields -= 1;
+                    spooled[spooledFields].accept(targetObject);
                 }
+            } catch (ParsingException e) {
+                throw new ParsingException(e.getLineNumber(), e.getColumnNumber(),
+                        "failed to build [" + objectParser.getName() + "] after last required field arrived", e);
+            } catch (Exception e) {
+                throw new ParsingException(null, "Failed to build [" + objectParser.getName() + "] after last required field arrived", e);
             }
         }
 
         /**
          * Spool up a consumer that we'll call once the targetObject is built. If targetObject has been built just apply the consumer
          * immediately.
          */
+        @SuppressWarnings("unchecked")
```

Comment:
That is what I did at first but I figured the method was so small it wan't worth the funky syntax dance. I can do it if you'd prefer though.